Adding Solutions and Code for Question 2 for CTCI on Feb. 7 (#175)
UWB-ACM/CTCI: 2020-01-Winter/3_stacks_queues/README.md

<!-- Don't remove -->
<a name="s2"/>

### 2. Recent Counter

Source: https://leetcode.com/problems/number-of-recent-calls/solution/

#### Naive/Simple Solution
**Time Complexity**: O(n) where N is the number of times `ping()` is called.

> **Space Complexity:** O(n) where N is the size of the window. In this case, 3000.

Reviewer: ```suggestion
**Space Complexity:** O(n) where N is the number of times `ping()` is called. In this case, the maximum size of the queue is bounded at 3001 elements.
```

Java Solution:
```java
class RecentCounter{
    //This is one way of making a FIFO Queue in Java
    Queue<Integer> queue = new LinkedList<>();

    //The ping function.
    //Calling this function requires passing in the time t in milliseconds
    public int ping(int t){
        //First, we eneueue the new time into the end of the queue
        queue.add(t);

        //Now we loop through the queue and dequeue from the front
        //any pings that happened more that 3000 seconds ago
        for(int i = 0; i < queue.size(); i++){
            if(queue.peek() < t-3000){
                queue.remove();
            }
        }

        //Since the problem statement wants the number of pings in
        //the previous 3000 milliseconds, we can now simply return
        //the number of elements in the queue
        return queue.size();
    }
}
```

Python Solution:
```python
class RecentCounter:
    #This queue is a Python list
    #In this problem, we treat it like a FIFO Queue

>     queue = []

Reviewer: This is fine for a single example, but the variable `queue` is being treated as static to the class instead of specific to the instance. If I try to add new examples / class instances, the contents of the variable are retained.
```suggestion
    def __init__(self):
        self.queue = []
```

    #The ping function. The self is implicit, so calling
    #this function requires passing in the time t in milliseconds
    #self is like "this" in Java
    def ping(self, t):
        #First, we eneueue the new time into the end of the queue
        self.queue.append(t)

        #Now we loop through the queue and dequeue from the front
        #any pings that happened more that 3000 seconds ago

>         for t in self.queue:

Reviewer: I think that the `for` loop is breaking when items are being removed; I don't remember the specifics of how iterables work when the corresponding data structure is modified. 

This homebrewed example should have multiple items being removed in the last few operations, but the queue retains more items than it should.

```
Num of pings at t = 0 :  1 ; queue is  [0]
Num of pings at t = 1 :  2 ; queue is  [0, 1]
Num of pings at t = 2 :  3 ; queue is  [0, 1, 2]
Num of pings at t = 3 :  4 ; queue is  [0, 1, 2, 3]
Num of pings at t = 4 :  5 ; queue is  [0, 1, 2, 3, 4]
Num of pings at t = 100 :  6 ; queue is  [0, 1, 2, 3, 4, 100]
Num of pings at t = 1000 :  7 ; queue is  [0, 1, 2, 3, 4, 100, 1000]
Num of pings at t = 2000 :  8 ; queue is  [0, 1, 2, 3, 4, 100, 1000, 2000]
Num of pings at t = 2500 :  9 ; queue is  [0, 1, 2, 3, 4, 100, 1000, 2000, 2500]
Num of pings at t = 3000 :  10 ; queue is  [0, 1, 2, 3, 4, 100, 1000, 2000, 2500, 3000]
Num of pings at t = 3003 :  10 ; queue is  [1, 2, 3, 4, 100, 1000, 2000, 2500, 3000, 3003]
Num of pings at t = 3005 :  10 ; queue is  [2, 3, 4, 100, 1000, 2000, 2500, 3000, 3003, 3005]
Num of pings at t = 4000 :  9 ; queue is  [4, 100, 1000, 2000, 2500, 3000, 3003, 3005, 4000]
```

```suggestion
        while self.queue and self.queue[0] < t - 3000:
```

Applying the suggestion would let you remove the next line as well, since it would be redundant.

Author: It could also be that my driver loop is far too narrow. I'll explore this

            if(self.queue[0] < t-3000):
                self.queue.pop(0)

        #Since the problem statement wants the number of pings in
        #the previous 3000 milliseconds, we can now simply return
        #the number of elements in the queue
        return len(self.queue)
```

#### Testing The Solutions OR Driver For Solution

> [Solutions available in the repository.]

Reviewer: Use hyperlinking syntax for links below
```suggestion
```

> **Java Solution:** (./3_stacks_queues/recent_counter/recent_counter_java.java)

Reviewer: ```suggestion
[**Java Solution**](./3_stacks_queues/recent_counter/recent_counter_java.java)
```

> **Python Solution:** (./3_stacks_queues/recent_counter/recent_counter_python.py)

Reviewer: ```suggestion
[**Python Solution**](./3_stacks_queues/recent_counter/recent_counter_python.py)
```
